File: codes/Shuofei/week 4/64_Minimum Path Sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        if ((grid.length == 0) && (grid[0].length == 0)) return 0;
        int[][] f = new int[grid.length][grid[0].length];
        for (int i = 0; i < grid.length; i++)
            for (int j = 0; j < grid[0].length; j++) {
                if ((i == 0) && (j == 0)) 
                    f[i][j] = grid[i][j];
                else if (i == 0)
                    f[i][j] = f[i][j-1] + grid[i][j];
                else if (j == 0)
                    f[i][j] = f[i-1][j] + grid[i][j];                
                else
                    f[i][j] = Math.min(f[i-1][j], f[i][j-1]) + grid[i][j];
            }
        return f[grid.length - 1][grid[0].length - 1];
    }
}

File: codes/Aiamjay/Week4-Day3/64. Minimum Path Sum.py
# encoding = utf-8
from math import inf


class Solution:
    def minPathSum(self, grid) -> int:
        num_row = len(grid)
        num_col = len(grid[0])
        if num_col == 0:
            return 0
        dp = [[0] * num_col for _ in range(num_row)]
        dp[0][0] = grid[0][0]
        for col in range(1, num_col):
            dp[0][col] = dp[0][col - 1] + grid[0][col]
        for row in range(1, num_row):
            dp[row][0] = dp[row - 1][0] + grid[row][0]
        for row in range(1, num_row):
            for col in range(1, num_col):
                dp[row][col] = min(dp[row][col - 1], dp[row - 1][col]) + grid[row][col]
        for row in dp:
            print(row)
        return dp[-1][-1]

    def test_solution(self):
        grid = [
            [1, 3, 1],
            [1, 5, 1],
            [4, 2, 1]
        ]
        print(self.minPathSum(grid))

        grid = [[1, 2, 3]]
        print(self.minPathSum(grid))
        pass


if __name__ == '__main__':
    s = Solution()
    s.test_solution()
